presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java

```diff
@@ -386,6 +438,239 @@ else if (exchange.getType() == ExchangeNode.Type.REPARTITION) {
+        {
+            if (!constantExpressions.isEmpty()) {
+                List<Symbol> constantSymbols = ImmutableList.copyOf(constantExpressions.keySet());
+
+                // update outputs
+                outputs = ImmutableList.<Symbol>builder()
+                        .addAll(outputs)
+                        .addAll(constantSymbols)
+                        .build();
+
+                // update inputs
+                inputs = inputs.stream()
+                        .map(input -> ImmutableList.<Symbol>builder()
+                                .addAll(input)
+                                .addAll(constantSymbols)
+                                .build())
+                        .collect(toImmutableList());
+
+                // update sources
+                sources = sources.stream()
+                        .map(source -> {
+                            Assignments.Builder assignments = Assignments.builder();
+                            assignments.putIdentities(source.getOutputSymbols());
+                            constantSymbols.forEach(symbol -> assignments.put(symbol, constantExpressions.get(symbol)));
+                            return new ProjectNode(idAllocator.getNextId(), source, assignments.build());
+                        })
+                        .collect(toImmutableList());
+            }
+
+            NewTableLayout insertLayout = metadata.getInsertLayout(session, tableHandle)
+                    // TODO: support insert into non partitioned table
+                    .orElseThrow(() -> new IllegalArgumentException("insertLayout for the temporary table must be present"));
+
+            PartitioningHandle partitioningHandle = expectedPartitioningMetadata.getPartitioningHandle();
+            List<String> partitionColumns = expectedPartitioningMetadata.getPartitionColumns();
+            ConnectorNewTableLayout expectedNewTableLayout = new ConnectorNewTableLayout(partitioningHandle.getConnectorHandle(), partitionColumns);
+            verify(insertLayout.getLayout().equals(expectedNewTableLayout), "unexpected new table layout");
+
+            Map<String, Symbol> columnNameToSymbol = temporaryTableColumns.entrySet().stream()
+                    .collect(toImmutableMap(entry -> entry.getValue().getName(), Map.Entry::getKey));
+            List<Symbol> partitioningSymbols = partitionColumns.stream()
+                    .map(columnNameToSymbol::get)
+                    .collect(toImmutableList());
+
+            InsertTableHandle insertTableHandle = metadata.beginInsert(session, tableHandle);
+            List<String> outputColumnNames = outputs.stream()
+                    .map(temporaryTableColumns::get)
+                    .map(ColumnMetadata::getName)
+                    .collect(toImmutableList());
+
+            InsertHandle insertHandle = new InsertHandle(insertTableHandle, new SchemaTableName("temporary", "table"));
+
+            return new TableFinishNode(
```

Comment:
I would flatten this to make it more readable.  Curious what others think.

Reply:
I done this this way for a reason. Usually that's what we are trying to do to increase readability. So the plan structure is visually readable.

Reply:
@arhimondr : What do you mean by "flatten" ? :)

Reply:
I mean put all the nodes as local variables so they're read in reverse order.  Not sure how others think about it, but I found it difficult to read the other parameters that come after the child node.

Reply:
I see your point. I agree that there's always a trade-off. But i think it is more important to be able to easily decode the structure of complex plan tree than to be able to quickly find some single parameters.